### DISCUSSION TITLE
Unicorn socket / potential dos?

### DISCUSSION THREAD

**Discussion by cfrademan** [Kludex/uvicorn#1519] [Potential Issue] [answered]
Hi guys,

Perhaps I am missing something here. Short version / background we integrated unicorn into our framework which has its own process manager for very good reasons over gunicorn (ton of extra functionality needed for large scale platform).  We did this along with falcon. Its all asyncio in short using the httptools protocol implementation within unicorn...  we simply send unicorn a socket to listen on all works great.. 

Recently during testing our team members pointed out some issues... We have a specific view/responder that is intended to live forever and continuously send data using generator. Faclon calls the ASGI send() continuously.... Which is exactly what its suppose todo...  Until we realised that even if the tcp socket closes abruptly from the client, httptools implementation  handles everything well, but never informs the caller of send() that something has gone wrong.. so it endlessly runs forever, until we have thousands of these... Yes we need need to be able to stream data over HTTP this is nothing new here.

However look at this snippet of code starting at Line 418 in protocols/http/httptools_impl.py:

418         if self.disconnected:
419             return

whoops, it just short-circuits the send, but endlessly the send just keeps getting called as expected... 

My humble opinion is does the ASGI standard cover this? If not, should we be raising RunTimeError of some sort? Disconnected Exception?

We can safely catch it, and end the endless loop.

Please advice? Since now the socket is stuck in TCP_WAIT. It would be easy for someone to run a DOS against our services.

**Comment by cfrademan** [comment, marked as answer]
Apparently some frameworks will raise an exception  if you send on disconnect client session. However its not within the ASGI 2 specifications. I have implemented a watcher task on a proxy stream object to wait for a HTTP Disconnect event which then kills the endless loop. Not sure if this the best solution, but it seems that Falcon has implemented this natively for SSE (server side events) when using SSE.

**Reply by euri10** [reply]
are you sure about the lines mentioned ? seems like it is on 447 and 448 ?
can you provide a minimal example of a problematic app ?
